Add release notes for Metals v0.10.0 (#2529)
scalameta/metals: website/blog/2021-02-19-tungsten.md

  </tr>
  <tr>
    <td>Merged PRs</td>
    <td align="center">72</td>
  </tr>
    <tr>
    <td>Contributors</td>
    <td align="center">14</td>
  </tr>
  <tr>
    <td>Closed issues</td>
    <td align="center">30</td>
  </tr>
  <tr>
    <td>New features</td>
    <td align="center">10</td>
  </tr>
</tbody>
</table>

For full details: https://github.com/scalameta/metals/milestone/34?closed=1

Metals is a language server for Scala that works with VS Code, Vim, Emacs,
Sublime Text and Eclipse. Metals is developed at the
[Scala Center](https://scala.epfl.ch/) and [VirtusLab](https://virtuslab.com)
with the help from [Lunatech](https://lunatech.com) along with contributors from
the community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/), and
give Metals a try!

- [Scala 3] Correct parser dialect for Scala 3 sources.
- Type decorations for definitions.
- Code action to insert the inferred type.
- Configuration option for default Scala version.
- Code lens to easily discover copy worksheet command.
- Improved short names when importing.
- [Scala 3] Filtering and sorting for completions.
- [Scala 3] Basic shorter signature support for hover.

> - Run/test capabilities when using sbt via BSP.

Reviewer: ```suggestion
- Debug capability for run/test when using sbt via BSP.
```

In previous versions of Metals, the completions offered would just be a list of
what is available in the current scope. This however did not provide the best
possible experience to the users. Thanks to great work by
[@sellophane](https://github.com/sellophane) those initial results are now
filtered and sorted so that the most useful methods and values will be available
first.

## [Scala 3] Basic shorter signature support for hover.

Together with better completions we also included a better way to display types
on hover. Previously, we would always display the full name with the package,
which would mean that some methods could look pretty complicated for example:

`def method(first: alpha.beta.gamma.DType, second: alpha.beta.gamma.DType)`

Thanks to [@tanishiking](https://github.com/tanishiking) we no longer display
the long type for anything that is already imported into scope. So the above
method would look like:

`def method(first: DType, second: DType)`

This needs to be still improved by shortening some of the symbols that are not
imported into scope and we should apply the same optimization in other places
where we display the type.

> ## Run/test capabilities when using sbt via BSP.

Reviewer: IMO, the highlight should be on `Debug`
```suggestion
## Debug capability for run/test when using sbt via BSP.
```

Author: Changed, but made it a bit shorter!

Thanks to the great work of [@adpi2](https://github.com/adpi2) and
[@er1c](https://github.com/er1c) on the
[scala-debug-adapter](https://github.com/scalacenter/scala-debug-adapter) Metals
now support running and debugging when using sbt BSP. The experience should be

> exactly the same as when using Bloop, since the new library reuses already

Reviewer: ```suggestion
almost the same as when using Bloop, since the new library reuses already
```
There are some differences because of the sbt part, for instance https://github.com/scalacenter/scala-debug-adapter/issues/32

Author: Fixed

> existing Bloop implementation and will be a standard to use for all BSP server.

Reviewer: ```suggestion
existing Bloop implementation and will be a standard to use for BSP servers.
```
it's too early to tell if Mill will integrate it or not

Author: Fixed

features, fixing bugs and improving the Scala 3 support at least to the level of
Scala 2. This will pave the way towards the `1.0.0` release later this year.

<table>
<tbody>
  <tr>
    <td>Commits since last release</td>
    <td align="center">156</td>
  </tr>
  <tr>
    <td>Merged PRs</td>
    <td align="center">74</td>
  </tr>
    <tr>
    <td>Contributors</td>
    <td align="center">14</td>
  </tr>
  <tr>
    <td>Closed issues</td>
    <td align="center">30</td>
  </tr>
  <tr>
    <td>New features</td>
    <td align="center">10</td>
  </tr>
</tbody>
</table>

For full details: https://github.com/scalameta/metals/milestone/34?closed=1

Metals is a language server for Scala that works with VS Code, Vim, Emacs,
Sublime Text and Eclipse. Metals is developed at the
[Scala Center](https://scala.epfl.ch/) and [VirtusLab](https://virtuslab.com)
with the help from [Lunatech](https://lunatech.com) along with contributors from
the community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/), and
give Metals a try!

- Type decorations for definitions.
- Insert inferred type refactoring.
- Configuration option for the default Scala version.

> - Code lens to easily discover copy worksheet command.

Reviewer 2: ```suggestion
- Code lens to easily discover the copy worksheet command.
```

Author: Fixed

- [Scala 3] Document outline
- [Scala 3] Code folding
- [Scala 3] Formatting `|` in multiline string
- [Scala 3] Code action to convert to multiline string.

## Type decorations for definitions.

One of the most
[requested features](https://github.com/scalameta/metals/issues/707) for Metals
was the ability to show the inferred type for anything that does not have it
defined explicitly. This includes the type at:

- value and variable definitions

```scala
val a = 1
var b = 2
```

- method definitions

```scala
def method() = ""
```

- variables in pattern matches

```scala
list match {
  case aa :: tail =>
}
```

- lambda parameters

```scala
list.map(param => param.toInt)
```

- for comprehensions

```scala
for(i <- 0 to 10)`
```

The decoration is based on the information produced during the compilation, so
anything new will first need to be compiled to have the type shown.

![type-decorations](https://i.imgur.com/NHzB0M6.png)

This feature can easily be controlled by the previously available
`metals.showInferredType` setting and will currently work for clients that allow
virtual decoration to be inserted into existing text. Those clients currently

> include Visual Studio Code, Emacs and Sublime Text (after merging the

Reviewer 2: ```suggestion
include Visual Studio Code, Emacs and Sublime Text (after merging this
```

Author: Fixed

[PR](https://github.com/scalameta/metals-sublime/pull/36)).

## Insert inferred type refactoring.

Another highly requested feature and closely related to the type decorations is
the ability to automatically insert type annotation. This is achieved using a
new `Insert type annotations` and
`Insert type annotation into pattern definition` refactor code actions. Those
actions work in the same places as mentioned in the previous point - method,

> variables, lambdas etc. - but uses the presentation compiler instead of indexes.

Reviewer 2: ```suggestion
variables, lambdas etc. - but use the presentation compiler instead of indexes.
```

Author: Fixed

What it means is that this new feature will work even in some more problematic
cases when the file is not yet compiled or in files where type decorations are

> not yet supported like in for example worksheets.

Reviewer 2: ```suggestion
not yet supported, like in worksheets.
```

Author: Fixed

![inferred-type](https://i.imgur.com/ZT36mH7.gif)

There might be still some work to do when it comes to some more complicated
types, but it should work perfectly well for most daily use cases. Please do

> report any issues you might encounter, it will help us greatly to polish this

Reviewer 2: ```suggestion
report any issues you might encounter, as it will help us greatly to polish this
```

Author: Fixed

Previously, for any standalone files Metals would use the default Scala version
2.12.13. This changed in `v0.10.0` and can either be defined with a setting:

```json
"metals.fallbackScalaVersion" : "3.0.0-RC1"
```

or when the setting is empty or set to `automatic` Metals will infer the Scala
version automatically from the project, which will be the newest Scala version
used in the project.

## Code lens for copy worksheet command.

In the previous version `0.9.8` a new command was added to Metals that allows
clients to export the worksheet output for the users to easily share their
results. This however was not easy to discover for a large chunk of the users,
which is why the command is also now available by using the code lens at the top
of a worksheet file. This will work for any clients that support the
`copy-worksheet-output` command.

## Debug run/test when using sbt via BSP.

Thanks to the great work of [@adpi2](https://github.com/adpi2) and
[@er1c](https://github.com/er1c) on the
[scala-debug-adapter](https://github.com/scalacenter/scala-debug-adapter) Metals
now supports debugging when using sbt BSP. The experience should be almost the

> same as when using Bloop, since the new library reuses already existing Bloop

Reviewer 2: ```suggestion
same as when using Bloop, since the new library reuses the already existing Bloop
```

Author: Fixed

implementation and will be a standard to use for BSP servers.

> To use BSP users still need to use `metals.bsp-switch` command, which will take

Reviewer 2: ```suggestion
To use sbt as a BSP server users still need to use `metals.bsp-switch` command, which will take
```

Author: Fixed

would just be a list of what is available in the current scope. This however did
not provide the best possible experience to the users. Thanks to work by
[@sellophane](https://github.com/sellophane) those initial results are now
filtered and sorted so that the most useful methods and values will be available
first.

## [Scala 3] Basic shorter signature support for hover.

Together with better completions we also included a better way to display types
on hover. Previously, we would always display the full class name with the
package prefix, which would mean that some methods could look quite complicated
for example:

`def method(first: alpha.beta.gamma.DType, second: alpha.beta.gamma.DType)`

Thanks to [@tanishiking](https://github.com/tanishiking) we no longer display
the long type for anything that is already imported into scope. So the above
method would look like:

`def method(first: DType, second: DType)`

This needs to be further improved by shortening some of the symbols that are not
imported into scope and we still need to apply the same optimization in other
places where the type is displayed.

## [Scala 3] Parser based features for Scala 3

> Metals uses underneath a separate parser implementation that can be found in

Reviewer 2: ```suggestion
Under the hood Metals uses a separate parser implementation that can be found in
```

Author: Fixed

<table>
<tbody>
  <tr>
    <td>Commits since last release</td>
    <td align="center">156</td>
  </tr>
  <tr>
    <td>Merged PRs</td>
    <td align="center">74</td>
  </tr>
    <tr>
    <td>Contributors</td>
    <td align="center">14</td>
  </tr>
  <tr>
    <td>Closed issues</td>
    <td align="center">30</td>
  </tr>
  <tr>
    <td>New features</td>
    <td align="center">10</td>
  </tr>
</tbody>
</table>

For full details: https://github.com/scalameta/metals/milestone/34?closed=1

Metals is a language server for Scala that works with VS Code, Vim, Emacs,
Sublime Text and Eclipse. Metals is developed at the
[Scala Center](https://scala.epfl.ch/) and [VirtusLab](https://virtuslab.com)
with the help from [Lunatech](https://lunatech.com) along with contributors from
the community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/), and
give Metals a try!

- Type decorations for definitions.
- Insert inferred type refactoring.
- Configuration option for the default Scala version.
- Code lens to easily discover the copy worksheet command.
- Improved short names when importing.
- Debug run/test when using sbt via BSP.

> - Support for Scala 3.0.0-RC1

Author: Yeah, I didn't want to duplicate it, but I think it might be sensible to make it as explicit as possible. We've had numerous questions about it previously. Will add!

![type-decorations](https://i.imgur.com/NHzB0M6.png)

This features can easily be controlled by the previously available
`metals.showInferredType` setting.

## Code action to insert inferred type.

Another highly requested feature and closely related to the type decorations is
the ability to automatically insert type annotation. This is achieved using a
new `Insert type annotations` and
`Insert type annotation into pattern definition` code actions. Those actions
work in the same places as mentioned in the previous point - method, variables,
lambdas etc. - but uses the presentation compiler instead of indexes. What it
means is that this new feature will work even in some more problematic cases,
where the file does not yet compiled or saved and files that type decorations
are not yet supported like in for example worksheets.

![inferred-type](https://i.imgur.com/ZT36mH7.gif)

There might be still some work to do when it comes to some more complicated
types, but it should work perfectly well for most daily use cases. Please do
report any issues you might encounter, it will help us greatly to polish
printing the types properly.

> The feature will currently work for clients that allow virtual text to be

Reviewer 3: Is this feature really dependent on the ability to show decorations ? Isn't code actions support from the client enough ?

Author: Good catch :O That was supposed to be mentioned for decorations

new feature.

## Configuration option for default Scala version.

Previously, for any standalone files Metals would use the default Scala version
2.12.13. This changed in `v0.10.0` and can either be defined with a setting:

```json
"metals.fallbackScalaVersion" : "3.0.0-RC1"
```

or when the setting is empty or set to `automatic` Metals will infer the Scala

> version automatically from the project, which will be the newest Scala version

Reviewer 4: ```suggestion
version automatically from the project, which will be the highest Scala version
```

Author: Fixed

In previous versions of Metals, the completions offered by the language server
would just be a list of what is available in the current scope. This however did
not provide the best possible experience to the users. Thanks to work by
[@sellophane](https://github.com/sellophane) those initial results are now
filtered and sorted so that the most useful methods and values will be available
first.

## [Scala 3] Basic shorter signature support for hover.

Together with better completions we also included a better way to display types
on hover. Previously, we would always display the full class name with the
package prefix, which would mean that some methods could look quite complicated
for example:

`def method(first: alpha.beta.gamma.DType, second: alpha.beta.gamma.DType)`

Thanks to [@tanishiking](https://github.com/tanishiking) we no longer display
the long type for anything that is already imported into scope. So the above
method would look like:

`def method(first: DType, second: DType)`

This needs to be further improved by shortening some of the symbols that are not
imported into scope and we still need to apply the same optimization in other
places where the type is displayed.

> ## [Scala 3] Parser based features for Scala 3

Author: Fixed

the long type for anything that is already imported into scope. So the above
method would look like:

`def method(first: DType, second: DType)`

This needs to be further improved by shortening some of the symbols that are not
imported into scope and we still need to apply the same optimization in other
places where the type is displayed.

## [Scala 3] Parser-based features for Scala 3

Under the hood Metals uses a separate parser implementation that can be found in
[scalameta](https://github.com/scalameta/scalameta) and it supports numerous
different Scala versions. Thanks to the recent work on supporting Scala 3, we
now have a dialect that should parse the Scala 3 syntax correctly on par with
3.0.0-RC1 version.

This parser is used by Metals to provide a number of different features, that
should now work correctly for Scala 3:

- document outline which shows the map for the current source file and can also
  be searched, which is useful to locate a particular method or value in the
  current document.
- folding ranges which enables users to hide parts of their code.
- automatic management of pipes in multiline string.
- code action to convert simple string into multiline string.

> - go to definition in Scala 3 dependency sources.

Author: I added this point and the sentence below to mention the recently merged improvements around navigating dependency sources and generating semanticdb for Scala 3.
